File: src/diagram/sort/QuickSort.py
#! py -3


def quick_sort(array: list, low: int, high: int):
    if low < high:
        mid = _quick_pass(array, low, high)
        quick_sort(array, low, mid - 1)
        quick_sort(array, mid + 1, high)


def _quick_pass(array: list, low: int, high: int) -> int:
    guard = array[low]
    i = low + 1
    j = high
    while True:
        while array[i] < guard and i < high:
            i += 1
        while array[j] > guard:
            j -= 1
        if i >= j:
            break
        array[i], array[j] = array[j], array[i]

    array[low] = array[j]
    array[j] = guard

    return j


if __name__ == "__main__":
    # li = [62, 34, 343, 6, 72, 45, 42, 25, 82, 46, 27, 42, 72]
    li = [0,0,0,2,0,5]
    quick_sort(li, 0, len(li)-1)
    # _quick_pass(li, 0, len(li)-1)
    print(li)

File: src/diagram/sort/QuickSort.java
package diagram.sort;

import java.util.Arrays;

public class QuickSort {
    public static void sort(int[] nums) {
        quick(nums, 0, nums.length - 1);
    }

    private static void quick(int[] nums, int left, int right) {
        if (left < right) {
            int mid = quick_pass(nums, left, right);
            quick(nums, left, mid - 1);
            quick(nums, mid, right);
        }
    }

    private static int quick_pass(int[] nums, int left, int right) {
        int grd = nums[left];
        int l = left + 1, r = right;
        while (l < r) {
            while (l < r && nums[l] <= grd) l++;
            while (l < r && nums[r] >= grd) r--;
            nums[l] = nums[r] + (nums[r] = nums[l]) * 0;
        }
        if (nums[left] >= nums[l])
            nums[left] = nums[l] + (nums[l] = nums[left]) * 0;
        return l;
    }

    public static void main(String[] args) {
        int[] nums = new int[]{5, 1, 2, 6, 4, 7, 8, 9, 3, 2, 5, 8, 7, 4, 6, 2, 1, 5, 2};
        sort(nums);
        System.out.println(Arrays.toString(nums));
    }
}
